RHMAP-9974

TommyJ1994 opened the pull request:
**Motivation**
Users are unable to delete projects if the domain contains an environment that is pointing to an Openshift Enterprise 3 MBaaS target that is no longer reachable (Deleted for example).
The behavior in the studio is that all the contents of the project are deleted (ie apps and cloud apps) but after ~1 minute of waiting the user gets an error saying "Failed to delete project.". The on prem team recommended adding a message for 4.2 to describe why project deletion fails.

**Edit**: Resolved conflicts and fixed bug introduced by a previous PR.

fh-automatron commented:
### Pull Request Quality Guidelines

As we continue to focus on quality, here are a few things to keep in mind:

* The code changes conform to the contributing & code style guides for this component (if they exist)
* A component owner should review your change (we've tagged these above)
* Automated testing has been added or updated for this change (e.g. UART, Unit/Acceptance tests)
* If updating server component this PR should be deployed to a grid somewhere (via `fhcap-cli`) - please provide a link to this
* Ensure JIRA is up to date before & after merge (e.g. deployment notes, tester notes, release notes, fix version)
* Any SonarQube issues should be addressed where applicable (use your own judgement)

Please ping a component owner (see `fh contrib`) when your builds are complete, changes are deployed and you're ready for a review.
Fhcap PR will be generated automatically.
To deploy your changes to a cluster place comment in this PR with pattern [x][y] where x=deploy and y=clustername you wish to deploy. 
You can also deploy your change manually using fhcap-cli

TommyJ1994 commented:
Retest this please

TommyJ1994 commented:
Retest this please

fh-qube commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -173,6 +173,15 @@ function responseHandler(params, cb) {
     });
 
+    // MbaaS is unreachable
+    if (err && err.code !== "undefined" && err.code !== null && err.code === "ENOTFOUND" || err.code === "ETIMEDOUT") {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1067)

fh-qube commented:
SonarQube analysis reported 1 issue:
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 1 major

Watch the comments in this conversation to review them.

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -175,6 +175,15 @@ function responseHandler(params, cb) {
+        log.logger.error("Error : Status Code", err.code, err);
+        err = {
+          httpCode: 500,
+          message: "MBaaS environment is not reachable. Please make the environment available or delete it, and then retry the project delete."
```
Can't have a message specifically for project delete in this function as it handles all of the requests made to an mbaas (e.g. list a form, deploy a form.)

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -175,6 +175,15 @@ function responseHandler(params, cb) {
     if (err || (httpResponse && httpResponse.statusCode >= 400)) {
 
+      // MbaaS is unreachable
+      if (err.code !== undefined && err.code === "ENOTFOUND" || err.code === "ETIMEDOUT") {
```
I think the idea is to have a relevant error message for all errors returned from an MBaaS so I don't think

`err.code === "ENOTFOUND" || err.code === "ETIMEDOUT"` is required.

TommyJ1994 replied:
@nialldonnellyfh If i remove these checks, when should the "MBaaS environment is not reachable..." message be triggered to be shown? The err.code has the value ENOTFOUND/ETIMEDOUT when the mbaas is unreachable (httpResponse.statusCode is '200 ok' though..).

nialldonnellyfh replied:
Actually yes. I would say keep them. Good idea.

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -175,6 +175,15 @@ function responseHandler(params, cb) {
+      if (err && err.code !== undefined && err.code === "ENOTFOUND" || err.code === "ETIMEDOUT") {
+        log.logger.error("Error : Status Code", err.code, err);
+        err = {
+          httpCode: 500,
```
This should be `httpResponse.statusCode` instead of assuming 500

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -175,6 +175,15 @@ function responseHandler(params, cb) {
+        log.logger.error("Error : Status Code", err.code, err);
+        err = {
+          httpCode: 500,
+          message: "MBaaS environment is not reachable. Please make the environment available or delete it, and then retry the project delete."
```
This error message should be more generic.

Something like, "MBaaS environment is not reachable. Please make the environment available or delete it."

Would be good to also add the Environment ID to the message.

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -181,6 +181,12 @@ function responseHandler(params, cb) {
+      } else if (err.code !== undefined && (err.code === "ENOTFOUND" || err.code === "ETIMEDOUT")) {
+        err = {
+          httpCode: httpResponse.statusCode || err.code || 500,
+          message: "MBaaS environment is not reachable. Please make the environment available or delete it. Environment ID - " + params.envLabel
```
`params.envLabel` is the label of the environment, not the ID.

nialldonnellyfh commented:
Reviewed. Small comment above.

fh-qube commented:
SonarQube analysis reported 2 issues:
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [Function has a complexity of 24 which is greater than 20 authorized.](https://github.com/feedhenry/fh-mbaas-client/blob/6a5921395a6c829a3351dcba829ba1d4ef4100b6/lib/mbaasRequest/mbaasRequest.js#L155) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [Function has a complexity of 23 which is greater than 20 authorized.](https://github.com/feedhenry/fh-mbaas-client/blob/6a5921395a6c829a3351dcba829ba1d4ef4100b6/lib/mbaasRequest/mbaasRequest.js#L157) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)

TommyJ1994 commented:
@nialldonnellyfh Quick review? Resolved conflicts from a PR merged yesterday and fixed a bug introduced by it. I can only get the environment label, not the ID so i just updated the text to say env label rather than env id. Will take a look at the sonarqube issues, but other than moving stuff into a separate function, i'm not too sure if the complexity can be cut down.

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -177,15 +177,20 @@ function responseHandler(params, cb) {
+      if (httpResponse && (httpResponse.statusCode || err.code)) {
         err = err || {};
-        err.httpCode = httpResponse.statusCode;
+        err.httpCode = httpResponse.statusCode || err.code;
```
`err.httpCode` is only for http Codes so I don't think `|| err.code` is required.

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -177,15 +177,20 @@ function responseHandler(params, cb) {
           };
+        } else if (err.code !== undefined && (err.code === "ENOTFOUND" || err.code === "ETIMEDOUT")) {
+          err = {
+            httpCode: httpResponse.statusCode || err.code || 500,
```
Same as above. The httpCode is for the numbers where `err.code` is a string.

TommyJ1994 replied:
Fixed.

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -177,15 +177,20 @@ function responseHandler(params, cb) {
             httpCode: 503,
             message: CONSTANTS.ERROR_MESSAGES.ENVIRONMENT_UNREACHABLE.replace('{{ENVLABEL}}', params.envLabel || "")
           };
+        } else if (err.code !== undefined && (err.code === "ENOTFOUND" || err.code === "ETIMEDOUT")) {
```
Safety check on err here would be good.

`else if (err && err.code .....`

TommyJ1994 replied:
Added. Good catch.

fh-qube commented:
SonarQube analysis reported 2 issues:
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [Function has a complexity of 23 which is greater than 20 authorized.](https://github.com/feedhenry/fh-mbaas-client/blob/5ccd001b43252d1bc9d3d60335505ce867375acb/lib/mbaasRequest/mbaasRequest.js#L155) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [Function has a complexity of 22 which is greater than 20 authorized.](https://github.com/feedhenry/fh-mbaas-client/blob/5ccd001b43252d1bc9d3d60335505ce867375acb/lib/mbaasRequest/mbaasRequest.js#L157) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)

nialldonnellyfh commented on lib/mbaasRequest/mbaasRequest.js:
```diff
@@ -177,7 +177,7 @@ function responseHandler(params, cb) {
 
       //There is a specific error message if an environment is unreachable
-      if (httpResponse && httpResponse.statusCode) {
+      if (httpResponse && (httpResponse.statusCode || err.code)) {
```
` || err.code` can be removed here also as it is not used in this case.

TommyJ1994 replied:
The `httpResponse.statusCode` that comes back from the mbaas for `"ENOTFOUND"/"ETIMEDOUT"` is undefined.  So it then resorts to the default error message and won't show the ."MBaaS environment is not reachable.." message.

nialldonnellyfh replied:
Can you switch it to `if (err && err.code) {` instead?

TommyJ1994 replied:
There is a `if (httpResponse.statusCode === 503) {` inside this if statement below, which could cause a problem. I think these if statements will need to be broken up.

nialldonnellyfh replied:
Yes, I think that's a good idea.

nialldonnellyfh commented:
Reviewed. Can you add a unit test to verify your changes also.

fh-qube commented:
SonarQube analysis reported 1 issue:
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [Function has a complexity of 21 which is greater than 20 authorized.](https://github.com/feedhenry/fh-mbaas-client/blob/68bd8d12fdbeeecebd59028cde78ee6c476854a5/lib/mbaasRequest/mbaasRequest.js#L155) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)

TommyJ1994 commented:
@nialldonnellyfh Added some unit tests.

nialldonnellyfh commented:
👍  Looks good.

fh-qube commented:
SonarQube analysis reported 1 issue:
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [Function has a complexity of 21 which is greater than 20 authorized.](https://github.com/feedhenry/fh-mbaas-client/blob/f0f93fd4877357908d1937ba0ba09ef47d40a920/lib/mbaasRequest/mbaasRequest.js#L158) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
